cmd/policy.go

```diff
@@ -70,42 +89,61 @@ func (sys *PolicySys) IsAllowed(args policy.Args) bool {
 	return args.IsOwner
 }
 
-// Init - initializes policy system from policy.json of all buckets.
-func (sys *PolicySys) Init(objAPI ObjectLayer) error {
-	if objAPI == nil {
-		return errInvalidArgument
-	}
-
+// Refresh in-memory bucket policy cache.
+func (sys *PolicySys) refreshInmemoryBucketpolicy(objAPI ObjectLayer) {
 	buckets, err := objAPI.ListBuckets(context.Background())
 	if err != nil {
-		return err
+		return
 	}
-
+	sys.RemoveDeletedBuckets(buckets)
```

Comment:
If you initialize `bucketPolicyMap` on load/reload, this method is not required.  What could be the problem if `bucketPolicyMap` is initialized?

Reply:
do you mean to do `bucketPolicyMap = make(....)` every 5 minutes? That would mean there will be a window of time where we will not have in-memory bucket policy and requests will not be honored with the bucket policy in that time

Reply:
Yep, but not the way you are referring.  Below is the method

```go
func (sys *PolicySys) reinit() {
	objAPI := newObjectLayerFn()
	if objAPI == nil {
		return
	}

	buckets, err := objAPI.ListBuckets(context.Background())
	if err != nil {
		return
	}

	bucketPolicyMap := make(map[string]policy.Policy)
	for _, bucket := range buckets {
		if config, err := GetPolicyConfig(objAPI, bucket.Name); err == nil {
			bucketPolicyMap[bucket.Name] = *config
		}
	}

	sys.Lock()
	sys.bucketPolicyMap = bucketPolicyMap
	sys.Unlock()
}
```

and `PolicySys.Init()` will start it as goroutine, before `return nil`
```
go func() {
	ticker := time.NewTicker(5 * time.Minute)
	defer ticker.Stop()

	for {
		select {
		case <-globalServiceDoneCh:
			break
		case <-ticker.C:
			sys.reinit()
		}
	}
}()
```

Reply:
No bala, there is more window of opportunity for race condition if we do it this way. Because reading policies of all buckets is going to take more time than reading policy of one bucket. Since RemoveDeletedBuckets() is a small function it is better to do it this way.

Reply:
I am unable to understand race condition here.   Could you elaborate about this?

Reading policies of all buckets is slow and there is no solution to it.  As disk is value of truth, we should reinit `PolicySys` than updating/modifying it.

Reply:
In the code below:
```
func (sys *PolicySys) reinit() {
	objAPI := newObjectLayerFn()
	if objAPI == nil {
		return
	}

	buckets, err := objAPI.ListBuckets(context.Background())
	if err != nil {
		return
	}

	bucketPolicyMap := make(map[string]policy.Policy)
	for _, bucket := range buckets {
		if config, err := GetPolicyConfig(objAPI, bucket.Name); err == nil {
			bucketPolicyMap[bucket.Name] = *config
		}
	}

	sys.Lock()
	sys.bucketPolicyMap = bucketPolicyMap
	sys.Unlock()
}
```

we do ListBuckets and read policy.json one by one, if during this process if the on disk policy is changed, we will not be able to pick up. In the current PR, the window for race condition is much lesser and hence better.